cpp/src/arrow/util/future.h

```diff
@@ -238,6 +242,22 @@ class Future {
     return impl_->Wait(seconds);
   }
 
+  template <typename... CancelArgs>
+  bool Cancel(CancelArgs&&... args) {
+    auto& stop_token = impl_->stop_token_;
```

Comment:
What if the future has already been marked completed?  This could happen with a timing issue (user requests cancel at the same time the operation is finishing) or when we mix this in with continuations it could happen where a chain of futures is cancelled from the end, the cancellation will propagate backwards through the chain, and the first few futures may have already completed.

This second case we could address when merging that change in too.

Reply:
That's a good question. In Java and Python, a completed Future cannot be cancelled (note the Python API is inspired from the Java API, AFAIK).